### ISSUE TITLE
\yii\db\QueryBuilder limit and offset rejects integer values in double type

### ISSUE THREAD

**Issue by m8rge** [yiisoft/yii2#7023]
PHP 5.6, Yii2 2.0.2

Ceil result passed to limit method will ignore value:
```php
\yii\db\ActiveRecord::find()->limit(ceil(5/2))->all();
```
Because ceil return type is double: 

> The return value of ceil() is still of type float as the value range of float is usually bigger than that of integer.

**Comment by gonimar** [comment]
Так же хорошо было бы разрешить `yii\db\Expression`

для выражений подобных этому:

```php
// PgSQL random record.
$model = Model::find()->offset(new Expression('random() * :count', [':count' => $count]))->limit(1)->one();
```

**Comment by cebe** [comment]
Related code:
- https://github.com/yiisoft/yii2/blob/master/framework/db/QueryBuilder.php#L806
- https://github.com/yiisoft/yii2/blob/master/framework/db/QueryBuilder.php#L826

**Comment by m8rge** [comment]
@gonimar хорошо бы не толкать все в один тикет =)

**Comment by samdark** [comment]
@gonimar your proposal won't be reviewed. Please create separate ticket since it's not realted.

**Comment by cebe** [comment]
@samdark it is related, limit and offset should not be limited to intergers but should allow float and expressions too.

**Comment by samdark** [comment]
Oh. OK. @gonimar sorry.

**Comment by dmitriy-krista** [comment]
Hi. I suggest these changes to prevent rejection integer values in double type (QueryBuilder.php):
```php
protected function hasLimit($limit)
    {
        return is_string($limit) && ctype_digit($limit) || is_integer($limit) && $limit >= 0 || is_float($limit) && ctype_digit($limit . '');
    }
```
```php
protected function hasOffset($offset)
    {
        return is_integer($offset) && $offset > 0 || is_string($offset) && ctype_digit($offset) && $offset !== '0' || is_float($offset) && ctype_digit($offset . '');
    }
```
By adding ...&& **is_float($limit) && ctype_digit($limit . '')** we can easily do like
```php
\yii\db\ActiveRecord::find()->limit(ceil(5/2))->all();
```
which will be equal
```php
\yii\db\ActiveRecord::find()->limit(3)->all();
```
 What do you think? If you agree with that I can make PR

**Comment by m8rge** [comment]
@dmitriy-krista I suggest cast any object to string:
```php
protected function hasLimit($limit)
    {
        return ctype_digit((string)$limit);
    }

**Comment by tino415** [comment]
Why not use is_numeric and accept almost anything changable to number ? https://github.com/yiisoft/yii2/pull/8268/files

**Comment by nineinchnick** [comment]
This might have been fixed by 5d4780d.

**Comment by m8rge** [comment]
Yes, you right!
fixed by https://github.com/yiisoft/yii2/commit/5d4780da219fbdd363b9df170442484e1ac5cabc
